text/3857-cfg-version.md

```diff
@@ -0,0 +1,634 @@
+- Allows evolution without requiring an MSRV bump
+- Its consistent with other predicates, e.g. `#[cfg(foo = "1.0")]`
+
+If we were stricter on the syntax,
+we could allow for version numbers to be directly accepted, without quotes 
+(e.g. `#[cfg(since(rust, 1.95.0))]`).
+If we ever decided to support operators (e.g.`#[cfg(since(rust, "=1.95.0"))]`, see `--check-cfg`), then we'd need to decide if those also go outside the string or then require a string, being inconsistent.
+This may limit us if we decide to allow for alternative version formats like with [target_version](#cfg_target_version) as they may not have formats that map well to SemVer.
+Worst case, we'd need to accept arbitrary bare words.
+This would also be inconsistent with other uses of `cfg`s
+*but* maybe that would just be the start to natively supporting more types in `cfg`,
+like integers which are of interest to embedded folks.
+
+## `--check-cfg`
+
+The `--check-cfg` predicate and the value for `rust` ensures users get warnings about
+- Invalid syntax
+- Using this with versions from before its supported, e.g. `#[cfg(since(rust, "1.0.0")]`
+
+`--check-cfg` requires a SemVer version, rather than a version requirement,
+in case we want the future possibility of relaxing SemVer versions
+*and* we want to infer from the fields used in `--check-cfg` to specify the maximum number of fields accepted in comparisons.
+
+We could have the `check-cfg` `since` predicate only apply to the `cfg` `since` predicate,
+causing `#[cfg(rust = "1.100.0")]` to warn.
+However,
+- the `since` predicates are a general feature intended to be used with other version numbers where exact matches may also be appropriate.
+- this would get in the way of approximating the vendor version by the language version for working around compiler bugs and snapshotting of compiler output.
+
+Possibly there could be a clippy lint specifically about `rust = "<something>"`.
+Alternatively, we could try to find a way to structure `--check-cfg` to allow the person defining the `cfg` to decide whether it can be used with `=` or not.
+One way of doing this is by allowing the `check-cfg` `since` predicate outside of the `values` predicate,
+meaning it works with the `cfg` `since` predicate and not the `=` operator.
+Another way would be for the `check-cfg` `since` predicate to never work with `=` but to instead
+allow operators inside of the `cfg` `since` predicate, e.g. `#[cfg(since(rust, "=1.95.0"))]`.
+However, with the rename of the predicate from `version` to `since`, operators don't fit in as easily.
+
+`--check-cfg` will cause the following to warn:
+```rust
+fn is_stderr_terminal() -> bool {
+    #[cfg(rust)]
+    #[cfg(since(rust, "1.70"))]
+    use std::io::IsTerminal as _;
+    #[cfg(rust)]
+    #[cfg(not(since(rust, "1.70")))]
+    use is_terminal::IsTerminal as _;
+    #[cfg(not(rust))]
+    use is_terminal::IsTerminal as _;
+
+    std::io::stderr().is_terminal()
+}
+```
+
```

Comment:
That's a reason for making `values(since(..))` only apply to the `since()` predicate (in `#[cfg]`), otherwise it becomes ambiguous when `--check-cfg` would warn.

Reply:
Unsure what this is referring to

Reply:
I'm not seeing how `values(none())` and `since()` not applying to `values()`
- are related
- causes ambiguity